### PR TITLE
Try to fix the potential data race in InformatonProviderImpl

### PR DESCRIPTION
**Pull request by PetrGuan**
I'm from Outlook iOS team. Not sure if this is a false positive, so please correct me if I'm wrong. The `Thread Sanitizer` complains the potential data race.

Screen Shot 2021-12-07 at 5 32 54 PM

### CODE REVIEW

File: lib/pal/InformationProviderImpl.cpp
Comment (lalitb) [inline lib/pal/InformationProviderImpl.cpp:76]: While this rightly increases the scope of mutex lock, it looks like it's not required to keep lock explicitly during callback function call (`callback::OnChanged`)

Comment (PetrGuan) [reply lib/pal/InformationProviderImpl.cpp:76]: Thanks for the comments! Any suggestions to this? Can we unlock?

Comment (lalitb) [reply lib/pal/InformationProviderImpl.cpp:76]: Don't see any straightforward fix. Either 
 - move the`local_callbacks` vector definition outside the lock guard. or
 - Use explicit lock/unlock instead of lock_guard.

Comment (lalitb) [reply lib/pal/InformationProviderImpl.cpp]: In general fine with it. Bringing `local_callbacks` vector outside may have some memory overhead in terms of heap allocation (std::allocator) every time the `InformatonProviderImpl::OnChanged` is called. But this should be negligible as most of the STL implementations have initial reserve capacity as `0` for std::vector. Still would be good if we can have another eye on this.

**Review by lalitb — APPROVED** [review]

**Review by lalitb — APPROVED** [review]
LGTM. Would be good if someone else too can review on the potential memory overhead of vector initialization.

File: lib/pal/InformationProviderImpl.cpp
Comment (lalitb) [reply lib/pal/InformationProviderImpl.cpp]: We can wait for a couple of days for comments and merge.

**Comment by lalitb** [comment]
I think it should be good to merge, don't see any further comments/concerns.

**Comment by PetrGuan** [comment]
Thanks @lalitb, appreciate your help ! Can you help me merge this? I'm not sure if i have the merge permission.

**Pull request merged**
